src/throttler/throttler.ts

```diff
@@ -254,4 +232,73 @@ export abstract class Throttler<T, R> {
       return p.resolve();
     });
   }
+
+  private initializeTimeout(taskData: TaskData<T, R>, cursorIndex: number): Promise<void> {
+    const timeoutMillis = taskData.timeoutMillis!;
+    const timeoutPromise = new Promise<void>((_, reject) => {
+      taskData.timeoutId = setTimeout(() => {
+        taskData.isTimedOut = true;
+        reject(new TimeoutError(this.taskName(cursorIndex), timeoutMillis));
+      }, timeoutMillis);
+    });
+
+    return timeoutPromise;
+  }
+
+  private async executeTask(taskData: TaskData<T, R>, cursorIndex: number): Promise<any> {
+    const t0 = Date.now();
+    try {
+      const result = await this.handler(taskData.task);
+      const dt = Date.now() - t0;
+      this.min = Math.min(dt, this.min);
+      this.max = Math.max(dt, this.max);
+      this.avg = (this.avg * this.complete + dt) / (this.complete + 1);
+
+      this.success++;
+      this.complete++;
+      this.active--;
+      return result;
+    } catch (err) {
+      if (taskData.retryCount === this.retries) {
+        throw new RetriesExhaustedError(this.taskName(cursorIndex), this.retries, err);
+      }
+      await backoff(taskData.retryCount + 1, this.backoff);
+      if (taskData.isTimedOut) {
+        throw new TimeoutError(this.taskName(cursorIndex), taskData.timeoutMillis!);
+      }
+      this.retried++;
+      taskData.retryCount++;
+      logger.debug(`[${this.name}] Retrying task`, this.taskName(cursorIndex));
+      return this.executeTask(taskData, cursorIndex);
+    }
+  }
+
+  private onTaskFulfilled(result: any, taskData: TaskData<T, R>, cursorIndex: number): void {
```

Comment:
Is it garanteed that when `onTaskFulfilled` is called, the task has not already timed out?
Consider the following sequences of `Promises` resolutions:
1. `executeTask` resolved
2. `timeoutPromise` promise rejected
3. `await Promise.race(promises)` is resolved successfully, but `isTimedOut` is already set to `true`.

I don't think this edge case is visible externally, but it is just confusing to have `isTimedOut` set to true for some successful tasks.

Perhaps have a task state so when a task timeout right after it gets fulfilled, it gets ignored. In that case, unregistering the timeout handle shouldn't be a critical step.
```
enum TaskState {
  PENDING,
  RUNNING,
  TIMEDOUT,
  FULLFILED
}
```

Reply:
I'm not sure if I'm understanding your concern correctly, so let me know if it's something other than this:
If `executeTask` resolves immediately before `timeoutPromise` is rejected, I believe node will place the `timeoutPromise`'s rejection code in the event loop *after* `await Promise.race(promises)` is resolved, stopping this from being an issue.

Reply:
If you believe that order is guaranteed, then I am fine with the current implementation.

However, it concerns me a bit that we need to think hard about these orders to prove its correctness (and could lead to bug if someone less unfamiliar add some feature). 

I can live with it, but it can be done in more readable way.

Reply:
* I don't see much benefit of differentiating the `PENDING` state from the rest
* `taskData.isTimedOut` differentiates running vs timed out
* Tasks that are fulfilled or timed out are removed from the taskData array, so `FULFILLED` wouldn't be useful

Even if we did use a TaskState enum, we'd still have to think about the problem the same way to prove its correctness. If a race condition did exist, it would exist regardless of the TaskState enum.

Reply:
In the timeout path, we check state is `PENDING` before setting to `TIMEDOUT`.
In the fulfilled path, we check state is `PENDING` before setting to `FULFILLED`.

There are two possible state transitions, and no possible racing condition could exist.
Canceling the `setTimeout` becomes optional.

Reply:
I'm not against adding the `TaskState` enum. However canceling timeout is pretty much a must since Node does not cancel the timeout counter in the background and it will block the CLI to exit

Reply:
There are no race conditions in your model or Tri's current model. Cancelling/clearing the `setTimeout` is already optional - It was only done to avoid the case where the timeout's promise not being resolved was the only thing keeping a node program from exiting.

Reply:
Agree. There is no race condition.

My comment is about simplicity and readability. If the code depends on JS runtime's priority around micro-tasks vs timers, it is not readable.

If you feel strongly against a state enum, don't block on this.